README
README - Net::SSLeay Perl module for using OpenSSL

13.6.2003, Sampo Kellomaki <sampo@symlabs.com>
Version: 1.16
$Id: README 53 2006-03-02 04:21:16Z mikem-guest $

1.23: tested against openssl-0.9.6j and openssl-0.9.7b
...
1.16: tested against openssl-0.9.6d, added more callback stuff
1.14: added perr certificate return, HPUX aCC flags fix
1.09: added many utlility functions of OpenSSL API, better old perl support
1.08: 64 bit fixes, windows fixes, fixed extra newline with auth bug
1.07: added rudimentary TLSv1 support by Stephen C. Koehler
1.06: fixed ssl_read_all() bug where input '0' would cause loop to exit
1.05: fixed certificate gen at make test
1.04: overhaul for OpenSSL-0.9.3b (try http://www.openssl.org/)

By popular demand...
--------------------

   perl -MNet::SSLeay -e '($p)=Net::SSLeay::get_https("www.openssl.org", 443, "/"); print $p'

Prerequisites
-------------

perl-5.6.1
                though anything starting from perl5.003 probably works.
OpenSSL-0.9.6j or OpenSSL-0.9.7b
                (try http://www.openssl.org/) - 
                This release has been tested with 0.9.6d and
		in historical light it seems likely that future versions
	  	will work as well (if major version number changes all bets
		are off, though)

Note: SSLeay is no longer supported. If you want to use Net::SSLeay with
      SSLeay or early versions of OpenSSL, use version 1.03. The support
      for SSLeay was dropped due to nobody maintaining it (all active
      work goes on with OpenSSL) and due to incompatible API changes
      in OpenSSL-0.9.2b. OpenSSL-0.9.1c support has also been dropped,
      version 1.03 was the last one to support that.

You should use the same C compiler and options to compile OpenSSL,
perl, and Net::SSLeay. This is the only supported configuration.
If you insist on using different compilers (perhaps because you
obtained either OpenSSL or perl as binaries from a vendor and they
used a compiler that you do not have) then all requests for support
will be ignored. If the only way for you to use the same compiler
for all three components is to recompile your openssl or perl, then
that is exactly what I expect you to do before asking for support.

Installing
----------

Unix:
	# build OpenSSL as per instructions in that package

	gunzip <Net_SSLeay.pm-1.06.tar.gz | tar xvf -
	cd Net_SSLeay.pm-1.14
	./Makefile.PL -t     # builds and tests it
	make install         # You probably have to su to root to do this
	perldoc Net::SSLeay  # optional, but highly recommended
	perldoc Net::SSLeay::Handle

HPUX:
	In principle the Unix build should work (Makefile.PL contains
	special code to detect aCC), but historically there have been
	some problems. Marko Asplund (aspa@@kronodoc._fi) reports
	that he has successfully compiled on HP-UX. He used following
	incantations

	Configuring OpenSSL:

		./Configure no-asm --prefix=/openssl/path hpux-parisc2-cc

	Configuring Net::SSLeay:
	
		perl Makefile.PL -path /openssl/path CCFLAGS='-D_HPUX_SOURCE \
		 -Aa -I/usr/local/include +e'

	The magic bit seemed to be the `+e' flag. Since version 1.14
	Makefile.PL tries to figure this out.

	He was using: gcc v2.95.2, OpenSSL v0.9.6c, Net::SSLeay-1.13

Windows:
	You need to get MS VC++ 6.0 Enterprise Edition. Nobody has
	reported building using CygWin, although I suspect the
	Unix instructions are pretty close.

	Add:
		d:\openssl
		d:\openssl\bin
		d:\openssl\lib
		d:\openssl\include
		d:\openssl\include\openssl

	to system's environment PATH, assuming you installed your openssl
	in d:\openssl. YMMV

		perl Makefile.PL -path D:\OpenSSL

	On windows, Makefile.PL will automatically set up library and include
	paths for windows compile and will automatcially find OpenSSL if it
	exists in c:\OpenSSL. If your OPenSSL libraries and includes are
	elsewhere, use the -path flag. If it does not work, please tweak
	Makefile.PL and submit a patch.

		nmake
		nmake install

	*** windows details are still being worked out. If you manage
	to compile this with different development environments under
	Win32, please post the diffs/success reports to http://alioth.debian.org/projects/net-ssleay
	
	The current incarnation of Windows tweaks was contributed by
	Eric A Selber <eselber@briefcase.com>

You should also be able to use CPAN.pm to install this module if you like.

Linking with RSAref is no longer supported (the patent issue is moot
doe to patent expiring). If you want to try it, you are on your own,
but here's how it used to work...

  For linking against RSAref add -rsaref flag like this:

	./Makefile.PL -rsaref -t  # builds and tests it, link against RSAref

  You must previously have built OpenSSL with RSAref support (which
  implies first building rsaref itself), I use the RSAglue method. File
  librsaref.a must be found in one of the locations searched by linker
  (-L switches). Usually this means that you have to rename rsaref.a to
  librsaref.a and copy it to suitable directory, e.g. /usr/local/ssl/lib.

  N.B. AFAIK the patent that made using RSAref necessary has expired, so
  this should be nonissue by now.


Problems (read this before sending mail)
----------------------------------------

Please, do not send bug report before you have

  - compiled your OpenSSL yourself - don't copy binaries, please
  - compiled your perl yourself and with substantially same CFLAGS
    and same C compiler (say `which cc' or `which gcc') as your OpenSSL.
    This is especially applicable to link errors and shared
    library loading problems. Please do not even dream of
    copying a perl binary or installing perl binary from a package.
    Perl's idea of calling conventions has to match OpenSSL's and
    unfortunately both are quite advanced pieces of code
    (guru duel: Larry Wall vs. Eric Young :-) with dynamic loading
    and who knows what
  - compiled my module from source against correct perl (say `which perl'
    and check your path). Generally my module's build process will
    discover correct compiler and flags from `perl -V'
  - tried gcc, if your vendor cc fails

If you post a question or make a bug report, please remeber to mention

  - Your platform and OS version (i386 Linux, Sparc Solaris, etc) (uname -a)
  - On Linux, please report glibc version as well (ls -l /lib/libc*)
  - Net::SSLeay version (see tar ball)
  - OpenSSL version (`/usr/local/ssl/bin/openssl version')
  - ANSI C compiler brand and version (e.g. gcc -v)

If build fails,
  - three compiler warnings are known to be emitted (due to lack of const
    in some places), one of them indicates a fatal bug in callback handling,
    but as I have not yet sorted it out, you'll simply have to ignore it
  - if you installed OpenSSL from some distribution, try getting a fresh
    copy from www.openssl.org and recompiling and installing it yourself
  - make sure you are not being confused by the fact that OpenSSL-0.9.3
    changed the location of include files to /usr/local/ssl/include/openssl/*
    Consider deleting all old bogus headers
  - if using newer than supported OpenSSL, please downgrade to supported
    version to see if it makes difference
  - you must compile the module, perl, and openssl with the same C compiler
    and the same options. Use perl -V to check what options were used and
    recompile openssl and Net::SSLeay accordingly
  - never report bugs related to binary installs. First compile _yourself_
    perl, openssl and my module, always using the same compiler and
    compiler flags. Many distros are known to "know better" and thus
    cause problems for their users. I'm not very sympathetic to having
    to answer end user questions thus created.
  - send full output of `make clean; perl Makefile.PL -t'

If make test fails, please
  - one warning is known to be emitted between tests 4 and 5 (callback)
  - edit test.pl and set $trace=2
  - send full output of `make clean; perl Makefile.PL -t'
  - send contents of sslecho.log

If you have problems with a site, please
  - what site, what server software (including version and platform)
  - does it reproduce with s_client, try with something like

   echo 'GET /' | /usr/local/ssl/bin/openssl s_client -connect www.bacus.pt:443

  - does it reproduce with popular web browsers
  - play with Net::SSLeay::ssl_version (see top of SSLeay.pm)
  - does the site run exotic configuration, e.g. insisting on specific
    protocol version, limiting available ciphers, using nonstandard
    ciphers, weird authentication arrangements, etc.)
  - contact the owner of the server to see what the problem looks like
    in his end. He should be able to tell you the exact versions used
    and the error messages he is seeing in his log
  - if you ask me to check a site out, you are granting me permission
    to access that site and will pay all legal expenses to defend me
    in court as well as any remedies that may be granted to the site
    in case the site decides to sue me. You warrant that you are
    authorized to give me permission to access the site.
  - if you ask me to check a site, please send me a working URL and
    include any authentication credentials if needed. If your site
    is so confidential that you can not give me an URL, then do
    not ask me to debug your problems.

HP-UX is known to give some problems, please mail me or the mailing
list so we can get these problems straightened. Hint: it has to do
with dynamic loading. One user reports that adding `-lgcc' to EXTRALIBS
and LD_LOAD_LIBS in Makefile fixes the problem. I have not received any
confirmation whether this fix really works, but its worth a try. Another
bag of problems is people installing against binary distributed
perl and compiling the package with different cc or different options.
Genereally this will never work. Please compile _yourself_ your perl,
openssl, and the module, always with the same compiler and compiler flags.

Solaris 8 does not come standard with /dev/random or /dev/urandom, and the
'make test' assumes that some source of randomness is available. 'make test'
will fail on Solaris 8 if /dev/urandom is not available. The error message
seen with trace enabled will be "SSL_GET_NEW_SESSION:ssl session id callback
failed". In order to fix this, you must install Sun patch 112438-03 from
http://sunsolve.sun.com

#: unzip 112438-03.zip
#: patchadd ./112438-03
You will probably need to reboot your system:
#: reboot

I have a report (schinder@@pobox._com) of make test segfaulting on
Linux-PPC. This still needs to be investigated. No recent information
has been received.

It seems perl5.004 (at least some versions) has bad xsub compiler which
can make builds sometimes fail. Try upgrading to perl-5.6.1 first.

"Random number generator not seeded!!!" This warning indicates that
    randomize() was not able to read /dev/random or /dev/urandom, possibly
    because your system does not have them or they are differently
    named. You can still use SSL, but the encryption will not be as
    strong.

Did you read the POD documentation (if you don't know what that
is, just say `perldoc Net::SSLeay' or `more SSLeay.pm')?

Are you sure you didn't confuse `Net::SSLeay' with `SSLeay' that
comes with OpenSSL?

My development environments used to be

  i686,   Linux-2.4.3,  gcc-2.92.2,   glibc-2.2,   perl-5.6.0,   openssl-0.9.6a
  i686,   Linux-2.4.3,  gcc-2.92.2,   glibc-2.2,   perl5.005_02, openssl-0.9.6a
  i686,   Linux-2.0.35, gcc-2.7.2.3,  glibc-2.0.6, perl5.005_02, openssl-0.9.5a
  i586,   Linux-2.4.3,  gcc-2.92.2.1, glibc-2.2.2, perl-5.6.0,   openssl-0.9.6a
  i586,   Linux-2.4.3,  gcc-2.92.2.1, glibc-2.2.2, perl5.005_03, openssl-0.9.6
  i586,   Linux-2.4.3,  gcc-2.92.2.1, glibc-2.2.2, perl5.005_03, openssl-0.9.6a
  Sun-U1, SunOS-5.6,    gcc-2.92.2,   libc2        perl-5.6.1,   openssl-0.9.6c

Unfortunately I do not have access to other systems so you are
somewhat on your own. Everything compiles without a warning (except
those mentioned above) on my systems.

Check if there are any post release building hints in

	http://www.symlabs.com/Net_SSLeay/index.html

Check that perl is finding your OpenSSL.

If `make test' bombs, add following line to the test script that fails:

	$Net::SSLeay::trace = 2;

and see what happens. You may also have to edit test.pl to make sure
the debugging output gets printed.

If `make test' prints lots of `connect: Connection refused...' errors,
then sslecho.pl test server has died. It is supposed to be launched in
the beginning of test.pl, but can fail if, e.g. port 1212 is taken or
in TIMEWAIT state. Look also in ssleacho.log file for diagnostics.

If you are really low on memory and the 1 MB tests fail, edit value of
$mb variable in test.pl.

If you get core dump, build your perl for debugging (add -g to
ccflags, see INSTALL in perl distribution), build your SSLeay for
debugging as well, add -g flag to Makefile.PL:

	make clean
	perl Makefile.PL -g
	make static
	make test_static
	gdb perl core       # post mortem
	  > bt              # show stack trace
	gdb perl            # run live with debugging
	  # set break point in SSLeay.xs or in suspect function of OpenSSL
	  > br XS_Net__SSLeay_connect
	  > run yourscript.pl arg arg

For gdb'ing make sure gdb finds all the relevant source code. This
may mean that you must run perl and OpenSSL from the directories where
the respective makefiles build them.

You can also enable PR and PRN macros in SSLeay.xs and sprinkle
even some more around the code to figure out what's happening.

Some exotic configurations of perl may cause unstability: make sure
OpenSSL uses the same malloc as perl. Recompile perl without
threads. Try not using the PerlIO abstraction.

If you need to tweak build for some platform, please let me know
so I can fix it. Patches and gdb session dumps are also welcome.

License and Copying
-------------------

Copyright (c) 1996-2002 Sampo Kellomaki <sampo@symlabs.com>
Copyright (c) 2005 Florian Ragwitz <rafl@debian.org>
Copyright (c) 2005 Mike McCauley <mikem@open.com.au>
All Rights Reserved.

Distribution and use of this module is under the same terms as the
OpenSSL package itself (i.e. free, but mandatory attribution; NO
WARRANTY). Please consult LICENSE file in the root of the OpenSSL
distribution.

While the source distribution of this perl module does not contain
Eric's or OpenSSL's code, if you use this module you will use OpenSSL
library. Please give Eric and OpenSSL team credit (as required by
their licenses).

And remember, you, and nobody else but you, are responsible for
auditing this module and OpenSSL library for security problems,
backdoors, and general suitability for your application.

Recommended reading
-------------------

===> HTTP protocol specification. It applies 100% to HTTPS too and doing
password authentication is explained there. <===

If you are newbie interested in grabbing web pages from https servers,
please read HTTP documentation from http://www.w3c.org/ before asking trivial
questions. That document also covers the basic-auth FAQ (URLs like
http://user:pass@host). Do not ask questions about authentication before
consulting the HTTP specification. HTTPS is just HTTP in SSL transport.

If you are doing advanced stuff, and don't find documentation you need,
please try to extrapolate from OpenSSL documentation (which unfortunately
is quite sparse) and the source code.

If you run into build problems, especially regarding shared libraries,
check your perl documentation, especially the perlxtut(1) man page,
which gives excellent tutorial of the build process of XSUBs.

  perlxtut(1)
  perlxs(1)
  perlguts(1)
  perlcall(1)

Say `perldoc Net::SSLeay' _NOW_!

To download OpenSSL, try URL http://www.openssl.org/

Of related interest may be `http://www.symlabs.com/Net_SSLeay/smime.html'

Bug reports, patch submission, feature requests, subversion access to the latest 
source code etc can be obtained at 
http://alioth.debian.org/projects/net-ssleay

The developer mailing list (for people interested in contributin to the source code)
can be found at 
http://lists.alioth.debian.org/mailman/listinfo/net-ssleay-devel



Changes
Revision history for Perl extension Net::SSLeay.

1.47 2012-04-04
     Fixed overlong lines in pod, patch from Salvatore Bonaccorso, Debian Perl
     Group
     Fixed spelling errors in pod, patch from Salvatore Bonaccorso, Debian Perl
     Group
     Fixed extra "garbage" files in 1.46 tarball. Patch from kmx.
     Fixed incorrect fail reports on some 64 bit platforms. Patch from paul.
     Fix to avoid FAIL reports from cpantesters with missing openssl
     Use my_snprintf from ppport.h to prevent link failures with perl 5.8 and
     earlier when compiled with MSVC.

1.46 2012-04-03
     Fixed a problem reported by Atoomic: 
      When bootstrapping Net::SSleay ( with DynaLoader ) if you override the SIG{DIE} signal, using 
      Net::SSLeay will result in an error.
      Recreated META.yml, added META.yml to dist
      Fixed typo: the word "corresponding" was mis-spelled as "coresponding"
       throughout the POD. Patched by kmx.
      Updated META.yml to include repository and bugtracker
     Constants cleanup - removing non existing constants (perhaps from pre-0.9.6 era) - kmx
     Automatic constants.c generation via helper_script/regen_openssl_constants.pl - kmx
     Future changes in constants now under better control via
     t/local/21_constants.t - kmx
     Added missing new files
     Reordering @EXPORT_OK (constants first, functions next) - kmx
     Adding missing 51 constants to @EXPORT_OK + test to keep it in sync - kmx
     Instructions "howto add new constant" added to helper_script/regen_openssl_constants.pl - kmx
     NEWLY INTRODUCED CONSTANTS:
     - Net::SSLeay::ASN1_STRFLGS_ESC_CTRL
     - Net::SSLeay::ASN1_STRFLGS_ESC_MSB
     - Net::SSLeay::ASN1_STRFLGS_ESC_QUOTE
     - Net::SSLeay::ASN1_STRFLGS_RFC2253
     - Net::SSLeay::ERROR_WANT_ACCEPT
     - Net::SSLeay::EVP_PKS_DSA
     - Net::SSLeay::EVP_PKS_EC
     - Net::SSLeay::EVP_PKS_RSA
     - Net::SSLeay::EVP_PKT_ENC
     - Net::SSLeay::EVP_PKT_EXCH
     - Net::SSLeay::EVP_PKT_EXP
     - Net::SSLeay::EVP_PKT_SIGN
     - Net::SSLeay::EVP_PK_DH
     - Net::SSLeay::EVP_PK_DSA
     - Net::SSLeay::EVP_PK_EC
     - Net::SSLeay::EVP_PK_RSA
     - Net::SSLeay::MBSTRING_ASC
     - Net::SSLeay::MBSTRING_BMP
     - Net::SSLeay::MBSTRING_FLAG
     - Net::SSLeay::MBSTRING_UNIV
     - Net::SSLeay::MBSTRING_UTF8
     - Net::SSLeay::OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION
     - Net::SSLeay::OP_CISCO_ANYCONNECT
     - Net::SSLeay::OP_CRYPTOPRO_TLSEXT_BUG
     - Net::SSLeay::OP_LEGACY_SERVER_CONNECT
     - Net::SSLeay::OP_NO_TLSv1_1
     - Net::SSLeay::OP_NO_TLSv1_2
     - Net::SSLeay::OP_SINGLE_ECDH_USE
     - Net::SSLeay::OP_TLS_BLOCK_PADDING_BUG
     - Net::SSLeay::X509_V_FLAG_CHECK_SS_SIGNATURE
     - Net::SSLeay::X509_V_FLAG_EXTENDED_CRL_SUPPORT
     - Net::SSLeay::X509_V_FLAG_POLICY_MASK
     - Net::SSLeay::X509_V_FLAG_USE_DELTAS
     - Net::SSLeay::X509_V_OK
     - Net::SSLeay::XN_FLAG_COMPAT
     - Net::SSLeay::XN_FLAG_DN_REV
     - Net::SSLeay::XN_FLAG_DUMP_UNKNOWN_FIELDS
     - Net::SSLeay::XN_FLAG_FN_ALIGN
     - Net::SSLeay::XN_FLAG_FN_LN
     - Net::SSLeay::XN_FLAG_FN_MASK
     - Net::SSLeay::XN_FLAG_FN_NONE
     - Net::SSLeay::XN_FLAG_FN_OID
     - Net::SSLeay::XN_FLAG_FN_SN
     - Net::SSLeay::XN_FLAG_MULTILINE
     - Net::SSLeay::XN_FLAG_ONELINE
     - Net::SSLeay::XN_FLAG_RFC2253
     - Net::SSLeay::XN_FLAG_SEP_COMMA_PLUS
     - Net::SSLeay::XN_FLAG_SEP_CPLUS_SPC
     - Net::SSLeay::XN_FLAG_SEP_MASK
     - Net::SSLeay::XN_FLAG_SEP_MULTILINE
     - Net::SSLeay::XN_FLAG_SEP_SPLUS_SPC
     - Net::SSLeay::XN_FLAG_SPC_EQ
     A number of tests were present in svn, but missing from MANIFEST, and
        were therefore not included in the dist. Added.
     NEWLY INTRODUCED FUNCTIONS:
     - Net::SSLeay::ASN1_INTEGER_free
     - Net::SSLeay::ASN1_INTEGER_get
     - Net::SSLeay::ASN1_INTEGER_new
     - Net::SSLeay::ASN1_INTEGER_set
     - Net::SSLeay::EVP_PKEY_assign_RSA
     - Net::SSLeay::EVP_PKEY_bits
     - Net::SSLeay::EVP_PKEY_free
     - Net::SSLeay::EVP_PKEY_new
     - Net::SSLeay::EVP_PKEY_size
     - Net::SSLeay::EVP_get_cipherbyname
     - Net::SSLeay::OPENSSL_add_all_algorithms_conf
     - Net::SSLeay::OPENSSL_add_all_algorithms_noconf
     - Net::SSLeay::OpenSSL_add_all_algorithms
     - Net::SSLeay::PEM_get_string_PrivateKey
     - Net::SSLeay::PEM_get_string_X509_CRL
     - Net::SSLeay::PEM_get_string_X509_REQ
     - Net::SSLeay::PEM_read_bio_PrivateKey
     - Net::SSLeay::PEM_read_bio_X509
     - Net::SSLeay::PEM_read_bio_X509_REQ
     - Net::SSLeay::P_ASN1_INTEGER_get_dec
     - Net::SSLeay::P_ASN1_INTEGER_get_hex
     - Net::SSLeay::P_ASN1_INTEGER_set_dec
     - Net::SSLeay::P_ASN1_INTEGER_set_hex
     - Net::SSLeay::P_ASN1_STRING_get
     - Net::SSLeay::P_X509_CRL_add_revoked_serial_hex
     - Net::SSLeay::P_X509_CRL_get_serial
     - Net::SSLeay::P_X509_CRL_set_serial
     - Net::SSLeay::P_X509_REQ_add_extensions
     - Net::SSLeay::P_X509_REQ_get_attr
     - Net::SSLeay::P_X509_add_extensions
     - Net::SSLeay::P_X509_copy_extensions
     - Net::SSLeay::P_X509_get_crl_distribution_points
     - Net::SSLeay::P_X509_get_ext_key_usage
     - Net::SSLeay::P_X509_get_key_usage
     - Net::SSLeay::P_X509_get_netscape_cert_type
     - Net::SSLeay::P_X509_get_pubkey_alg
     - Net::SSLeay::P_X509_get_signature_alg
     - Net::SSLeay::P_PKCS12_load_file
     - Net::SSLeay::X509V3_EXT_print
     - Net::SSLeay::X509_CRL_digest
     - Net::SSLeay::X509_CRL_free
     - Net::SSLeay::X509_CRL_get_issuer
     - Net::SSLeay::X509_CRL_get_lastUpdate
     - Net::SSLeay::X509_CRL_get_nextUpdate
     - Net::SSLeay::X509_CRL_get_version
     - Net::SSLeay::X509_CRL_new
     - Net::SSLeay::X509_CRL_set_issuer_name
     - Net::SSLeay::X509_CRL_set_lastUpdate
     - Net::SSLeay::X509_CRL_set_nextUpdate
     - Net::SSLeay::X509_CRL_set_version
     - Net::SSLeay::X509_CRL_sign
     - Net::SSLeay::X509_CRL_sort
     - Net::SSLeay::X509_CRL_verify
     - Net::SSLeay::X509_EXTENSION_get_critical

inc/Test/Exception.pm
#line 1
use strict;
use warnings;

package Test::Exception;
use Test::Builder;
use Sub::Uplevel qw( uplevel );
use base qw( Exporter );
use Carp;

our $VERSION = '0.25';
our @EXPORT = qw(dies_ok lives_ok throws_ok lives_and);

my $Tester = Test::Builder->new;

sub import {
    my $self = shift;
    if ( @_ ) {
        my $package = caller;
        $Tester->exported_to( $package );
        $Tester->plan( @_ );
    };
    $self->export_to_level( 1, $self, $_ ) foreach @EXPORT;
}

#line 78


sub _try_as_caller {
    my $coderef = shift;
    eval { uplevel 3, $coderef };
    return $@;
};


sub _is_exception {
    my $exception = shift;
    return ref $exception || $exception ne '';
};


sub _exception_as_string {
    my ( $prefix, $exception ) = @_;
    return "$prefix normal exit" unless _is_exception( $exception );
    my $class = ref $exception;
    $exception = "$class ($exception)" 
            if $class && "$exception" !~ m/^\Q$class/;
    chomp $exception;
    return "$prefix $exception";
};


#line 125


sub dies_ok (&;$) {
    my ( $coderef, $description ) = @_;
    my $exception = _try_as_caller( $coderef );
    my $ok = $Tester->ok( _is_exception($exception), $description );
    $@ = $exception;
    return $ok;
}


#line 165

sub lives_ok (&;$) {
    my ( $coderef, $description ) = @_;
    my $exception = _try_as_caller( $coderef );
    my $ok = $Tester->ok( ! _is_exception( $exception ), $description );
	$Tester->diag( _exception_as_string( "died:", $exception ) ) unless $ok;
    $@ = $exception;
    return $ok;
}


#line 218


sub throws_ok (&$;$) {
    my ( $coderef, $expecting, $description ) = @_;
    croak "throws_ok: must pass exception class/object or regex" 
        unless defined $expecting;
    $description = _exception_as_string( "threw", $expecting )
    	unless defined $description;
    my $exception = _try_as_caller( $coderef );
    my $regex = $Tester->maybe_regex( $expecting );
    my $ok = $regex 
        ? ( $exception =~ m/$regex/ ) 
        : eval { 
            $exception->isa( ref $expecting ? ref $expecting : $expecting ) 
        };
    $Tester->ok( $ok, $description );
    unless ( $ok ) {
        $Tester->diag( _exception_as_string( "expecting:", $expecting ) );
        $Tester->diag( _exception_as_string( "found:", $exception ) );
    };
    $@ = $exception;
    return $ok;
};


#line 272

sub lives_and (&;$) {
    my ( $test, $description ) = @_;
    {
        local $Test::Builder::Level = $Test::Builder::Level + 1;
        my $ok = \&Test::Builder::ok;
        no warnings;
        local *Test::Builder::ok = sub {
            $_[2] = $description unless defined $_[2];
            $ok->(@_);
        };
        use warnings;
        eval { $test->() } and return 1;
    };
    my $exception = $@;
    if ( _is_exception( $exception ) ) {
        $Tester->ok( 0, $description );
        $Tester->diag( _exception_as_string( "died:", $exception ) );
    };
    $@ = $exception;
    return;
}

#line 431

1;


inc/Module/Install/External.pm
#line 1
package Module::Install::External;

# Provides dependency declarations for external non-Perl things

use strict;
use Module::Install::Base;

use vars qw{$VERSION $ISCORE @ISA};
BEGIN {
	$VERSION = '0.76';
	$ISCORE  = 1;
	@ISA     = qw{Module::Install::Base};
}

sub requires_external_cc {
	my $self = shift;

	# We need a C compiler, use the can_cc method for this
	unless ( $self->can_cc ) {
		print "Unresolvable missing external dependency.\n";
		print "This package requires a C compiler.\n";
		print STDERR "NA: Unable to build distribution on this platform.\n";
		exit(0);
	}

	# Unlike some of the other modules, while we need to specify a
	# C compiler as a dep, it needs to be a build-time dependency.

	1;
}

sub requires_external_bin {
	my ($self, $bin, $version) = @_;
	if ( $version ) {
		die "requires_external_bin does not support versions yet";
	}

	# Load the package containing can_run early,
	# to avoid breaking the message below.
	$self->load('can_run');

	# Locate the bin
	print "Locating required external dependency bin:$bin...";
	my $found_bin = $self->can_run( $bin );
	if ( $found_bin ) {
		print " found at $found_bin.\n";
	} else {
		print " missing.\n";
		print "Unresolvable missing external dependency.\n";
		print "Please install '$bin' seperately and try again.\n";
		print STDERR "NA: Unable to build distribution on this platform.\n";
		exit(0);
	}

	# Once we have some way to specify external deps, do it here.
	# In the mean time, continue as normal.

	1;
}

1;

__END__

#line 138


inc/Array/Compare.pm
#line 1
#
# $Id: Compare.pm 22 2007-04-01 15:03:22Z dave $
#

#line 177

package Array::Compare;

use strict;
use vars qw($VERSION $AUTOLOAD);

use Carp;

$VERSION = 1.14;

my %_defaults = (Sep => '^G',
		 WhiteSpace => 1,
                 Case => 1,
		 Skip => {},
		 DefFull => 0);

#line 236

sub new {
  my $class = shift;

  my $self = {%_defaults, @_};

  bless $self, $class;

  return $self;
}

#
# Utility function to check the arguments to any of the comparison
# function. Ensures that there are two arguments and that they are
# both arrays.
#
sub _check_args {
  my $self = shift;
  croak('Must compare two arrays.') unless @_ == 2;
  croak('Argument 1 is not an array') unless ref($_[0]) eq 'ARRAY';
  croak('Argument 2 is not an array') unless ref($_[1]) eq 'ARRAY';

  return;
}

#line 267

sub compare_len {
  my $self = shift;

  $self->_check_args(@_);

  return @{$_[0]} == @{$_[1]};
}

#line 287

sub compare {
  my $self = shift;

  if ($self->DefFull) {
    return $self->full_compare(@_);
  } else {
    return $self->simple_compare(@_);
  }
}

#line 309

sub simple_compare {
  my $self = shift;

  $self->_check_args(@_);

  my ($row1, $row2) = @_;

  # No point in continuing if the number of elements is different.
  return unless $self->compare_len(@_);

  # @check contains the indexes into the two arrays, i.e. the numbers
  # from 0 to one less than the number of elements.
  my @check = 0 .. $#$row1;

  my ($pkg, $caller) = (caller(1))[0, 3];
  my $perm = $caller eq __PACKAGE__ . "::perm";

  # Filter @check so it only contains indexes that should be compared.
  # N.B. Makes no sense to do this if we are called from 'perm'.
  unless ($perm) {
    @check = grep {!(exists $self->Skip->{$_}
		     && $self->Skip->{$_}) } @check
		       if keys %{$self->Skip};
  }

  # Build two strings by taking array slices containing only the columns
  # that we shouldn't skip and joining those array slices using the Sep
  # character. Hopefully we can then just do a string comparison.
  # Note: this makes the function liable to errors if your arrays
  # contain the separator character.
  my $str1 = join($self->Sep, @{$row1}[@check]);
  my $str2 = join($self->Sep, @{$row2}[@check]);

  # If whitespace isn't significant, collapse it
  unless ($self->WhiteSpace) {
    $str1 =~ s/\s+/ /g;
    $str2 =~ s/\s+/ /g;
  }

  # If case isn't significant, change to lower case
  unless ($self->Case) {
    $str1 = lc $str1;
    $str2 = lc $str2;
  }

  return $str1 eq $str2;
}

#line 376

sub full_compare {
  my $self = shift;

  $self->_check_args(@_);

  my ($row1, $row2) = @_;

  # No point in continuing if the number of elements is different.
  # Because of the expected return value from this function we can't
  # just say 'the arrays are different'. We need to do some work to
  # calculate a meaningful return value.
  # If we've been called in array context we return a list containing
  # the number of the columns that appear in the longer list and aren't
  # in the shorter list. If we've been called in scalar context we
  # return the difference in the lengths of the two lists.
  unless ($self->compare_len(@_)) {
    if (wantarray) {
      my ($max, $min);
      if ($#{$row1} > $#{$row2}) {
	($max, $min) = ($#{$row1}, $#{$row2} + 1);
      } else {
	($max, $min) = ($#{$row2}, $#{$row1} + 1);
      }
      return ($min .. $max);
    } else {
      return abs(@{$row1} - @{$row2});
    }
  }

  my ($arr1, $arr2) = @_;

  my @diffs = ();

  foreach (0 .. $#{$arr1}) {
    next if keys %{$self->Skip} && $self->Skip->{$_};

    my ($val1, $val2) = ($arr1->[$_], $arr2->[$_]);
    unless ($self->WhiteSpace) {
      $val1 =~ s/\s+/ /g;
      $val2 =~ s/\s+/ /g;
    }

    unless ($self->Case) {
      $val1 = lc $val1;
      $val2 = lc $val2;
    }

    push @diffs, $_ unless $val1 eq $val2;
  }

  return wantarray ? @diffs : scalar @diffs;
}

#line 441

sub perm {
  my $self = shift;

  return $self->simple_compare([sort @{$_[0]}], [sort @{$_[1]}]);
}

#
# Attempt to be clever with object attributes.
# Each object attribute is always accessed using an access method.
# None of these access methods exist in the object code.
# If an unknown method is called then the AUTOLOAD method is called
# in its place with the same parameters and the variable $AUTOLOAD
# set to the name of the unknown method.
#
# In this function we work out which method has been called and
# simulate it by returning the correct attribute value (and setting
# it to a new value if the method was passed a new value to use).
#
# We're also a little cleverer than that as we create a new method on
# the fly so that the next time we call the missing method it has
# magically sprung into existance, thereby avoiding the overhead of
# calling AUTOLOAD more than once for each method called.
#
sub AUTOLOAD {
  no strict 'refs';
  my ($self, $val) = @_;
  my ($name) = $AUTOLOAD =~ m/.*::(\w*)/;

  *{$AUTOLOAD} = sub { return @_ > 1 ?
			 $_[0]->{$name} = $_[1] :
			   $_[0]->{$name}};

  return defined $val ? $self->{$name} = $val : $self->{$name};
}

#
# One (small) downside of the AUTOLOAD trick, is that we need to
# explicitly define a DESTROY method to prevent Perl from passing
# those calls to AUTOLOAD. In this case we don't need to do anything.
#
sub DESTROY { }

1;
__END__

#line 503

#
# $Log$
# Revision 1.13  2005/09/21 09:23:40  dave
# Documentation fix
#
# Revision 1.12  2005/03/01 09:05:33  dave
# Changes to pass Pod::Coverage tests (and, hence, increase kwalitee)
#
# Revision 1.11  2004/10/23 08:11:32  dave
# Improved test coverage
#
# Revision 1.10  2004/10/22 20:32:48  dave
# Improved docs for full comparison
#
# Revision 1.9  2003/09/19 09:37:40  dave
# Bring CVS version into line with old file
#
# Revision 1.1  2003/09/19 09:34:43  dave
# Bit of an overhaul
#
# Revision 1.7  2002/03/29 17:45:09  dave
# Test version
#
# Revision 1.6  2002/01/09 11:41:52  dave
# Small cleanups
#
# Revision 1.5  2001/12/09 19:31:47  dave
# Cleanup.
#
# Revision 1.4  2001/06/04 20:47:01  dave
# RCS Import
#
# Revision 1.3  2001/02/26 13:34:41  dave
# Added case insensitivity.
#
# Revision 1.2  2000/06/04 17:43:14  dave
# Renamed 'manifest' and 'readme' to 'MANIFEST' and 'README'.
# Added header info.
#
# Revision 1.1.1.1  2000/06/04 17:40:19  dave
# CVS import
#
# Revision 0.2  00/05/13  14:23:48  14:23:48  dave (Dave Cross)
# Added 'perm' method.
# Revision 0.1  00/04/25  13:33:55  13:33:55  dave (Dave Cross)
# Initial version.
#
